File: MainProject/Scenes/HexPlanet/Util/HexMetrics.cs
using System;
using Godot;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Enum;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Struct;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Util.HexSphereGrid;

namespace ZeromaXsPlaygroundProject.Scenes.HexPlanet.Util;

public static class HexMetrics
{
    public const float StandardRadius = 150f; // 150f 半径时才是标准大小，其他时候需要按比例缩放
    public const float StandardDivisions = 10f; // 10 细分时才是标准大小，其他时候需要按比例缩放

    public const float OuterToInner = 0.8660254037f; // √3/2 = 0.8660254037f
    public const float InnerToOuter = 1f / OuterToInner;
    public const float SolidFactor = 0.8f;
    public const float BlendFactor = 1f - SolidFactor;

    private const int TerracesPerSlope = 2;
    public const int TerraceSteps = TerracesPerSlope * 2 + 1;
    private const float HorizontalTerraceStepSize = 1f / TerraceSteps;
    public const float VerticalTerraceStepSize = 1f / (TerracesPerSlope + 1);

    // 适用于球面的阶地 Lerp
    public static Vector3 TerraceLerp(Vector3 a, Vector3 b, int step)
    {
        var bWithAHeight = Math3dUtil.ProjectToSphere(b, a.Length());
        var h = step * HorizontalTerraceStepSize;
        var horizontal = a.Slerp(bWithAHeight, h);
        var v = ((step + 1) / 2) * VerticalTerraceStepSize;
        var vertical = Mathf.Lerp(a.Length(), b.Length(), v);
        return Math3dUtil.ProjectToSphere(horizontal, vertical);
    }

    public static Color TerraceLerp(Color a, Color b, int step)
    {
        var h = step * HorizontalTerraceStepSize;
        return a.Lerp(b, h);
    }

    public static HexEdgeType GetEdgeType(int elevation1, int elevation2)
    {
        if (elevation1 == elevation2)
            return HexEdgeType.Flat;
        return Mathf.Abs(elevation1 - elevation2) == 1 ? HexEdgeType.Slope : HexEdgeType.Cliff;
    }

    #region 河流与水面

    public const float StreamBedElevationOffset = -1.75f;
    public const float WaterElevationOffset = -0.5f;
    public const float WaterFactor = 0.6f;
    public const float WaterBlendFactor = 1f - WaterFactor;

    #endregion

    #region 特征

    public const float WallTowerThreshold = 0.6f;
    public const float BridgeDesignLength = 7f;

    #endregion
}

File: TerraObserver/TO.FSharp/Commons/Utils/HexMetrics.fs
namespace TO.FSharp.Commons.Utils

open Godot
open TO.FSharp.Commons.Enums.Tiles
open TO.FSharp.Commons.Structs.Planets

/// 六边形测量数据
/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH
module HexMetrics =
    let private terracesPerSlope = 2
    let terraceSteps = terracesPerSlope * 2 + 1
    let private horizontalTerracesStepSize = 1f / float32 terraceSteps
    let verticalTerraceStepSize = 1f / float32 (terracesPerSlope + 1)

    // 适用于球面的阶地 Lerp
    let terraceLerp (a: Vector3) b step =
        let bWithAHeight = Math3dUtil.ProjectToSphere(b, a.Length())
        let h = float32 step * horizontalTerracesStepSize
        let horizontal = a.Slerp(bWithAHeight, h)
        let v = float32 ((step + 1) / 2) * verticalTerraceStepSize
        let vertical = Mathf.Lerp(a.Length(), b.Length(), v)
        Math3dUtil.ProjectToSphere(horizontal, vertical)

    let terraceLerpEdgeV (a: EdgeVertices) (b: EdgeVertices) (step: int) =
        EdgeVertices(
            V1 = terraceLerp a.V1 b.V1 step,
            V2 = terraceLerp a.V2 b.V2 step,
            V3 = terraceLerp a.V3 b.V3 step,
            V4 = terraceLerp a.V4 b.V4 step,
            V5 = terraceLerp a.V5 b.V5 step
        )

    let terraceLerpColor (a: Color) (b: Color) (step: int) =
        let h = float32 step * horizontalTerracesStepSize
        a.Lerp(b, h)

    let getEdgeType (elevation1: int) (elevation2: int) =
        if elevation1 = elevation2 then
            HexEdgeType.Flat
        elif Mathf.Abs(elevation1 - elevation2) = 1 then
            HexEdgeType.Slope
        else
            HexEdgeType.Cliff

    // ========== [河流与水面] ==========
    let streamBedElevationOffset = -1.75f
    let waterElevationOffset = -0.5f
    let waterFactor = 0.6f
    let waterBlendFactor = 1f - waterFactor

    // ========== [特征] ==========
    let wallTowerThreshold = 0.6f
    let bridgeDesignLength = 7f

/// 六边形测量数据
/// 作为对其他 .NET 公开的 F# 库，需要遵循《F# 组件设计准则 - 命名空间和类型设计 - 使用命名空间、类型和成员作为组件的主要组织结构》
/// https://learn.microsoft.com/zh-cn/dotnet/fsharp/style-guide/component-design-guidelines#use-namespaces-types-and-members-as-the-primary-organizational-structure-for-your-components
[<AbstractClass; Sealed>]
type HexMetrics =
    static member StandardRadius = 150f // 150f 半径时才是标准大小，其他时候需要按比例缩放
    static member StandardDivisions = 10f // 10 细分时才是标准大小，其他时候需要按比例缩放

    static member OuterToInner = 0.8660254037f // √3/2 = 0.8660254037f
    static member InnerToOuter = 1f / HexMetrics.OuterToInner
    static member SolidFactor = 0.8f
    static member BlendFactor = 1f - HexMetrics.SolidFactor
